robot/impl/local_robot_test.go

```diff
@@ -3359,3 +3359,76 @@ func TestCloudMetadata(t *testing.T) {
 		})
 	})
 }
+
+func TestReconfigureOnModuleRename(t *testing.T) {
+	ctx := context.Background()
+	logger := logging.NewTestLogger(t)
+
+	// Precompile complex module to avoid timeout issues when building takes too long.
+	complexPath := rtestutils.BuildTempModule(t, "examples/customresources/demos/complexmodule")
+
+	// Manually define mybase model, as importing it can cause double registration.
+	myBaseModel := resource.NewModel("acme", "demo", "mybase")
+
+	// Create config with at least one module and a component from that module
+	cfg := &config.Config{
+		Modules: []config.Module{
+			{
+				Name:     "mod",
+				ExePath:  complexPath,
+				LogLevel: "info",
+			},
+		},
+		Components: []resource.Config{
+			{
+				Name:  "myBase",
+				API:   base.API,
+				Model: myBaseModel,
+				Attributes: rutils.AttributeMap{
+					"motorL": "motor1",
+					"motorR": "motor2",
+				},
+			},
+			{
+				Name:                "motor1",
+				API:                 motor.API,
+				Model:               fakeModel,
+				ConvertedAttributes: &fakemotor.Config{},
+			},
+			{
+				Name:                "motor2",
+				API:                 motor.API,
+				Model:               fakeModel,
+				ConvertedAttributes: &fakemotor.Config{},
+			},
+		},
+	}
+
+	// Create copy of cfg since Reconfigure (called when setting up a robot) modifies cfg.
+	cfgCopy := *cfg
+	r := setupLocalRobot(t, ctx, cfg, logger)
+
+	// Use copy to modify config by renaming module
+	cfgCopy.Modules[0].Name = "mod-renamed"
+
+	// Create copy of cfgCopy to test against
+	expectedCfg := cfgCopy
+	r.Reconfigure(ctx, &cfgCopy)
+
+	actualCfg := r.Config()
+
+	// Verify attributes of config
+
+	// Manually inspect components
+	test.That(t, len(actualCfg.Components), test.ShouldEqual, 3)
+	for _, comp := range actualCfg.Components {
+		isMyBase := comp.Equals(cfg.Components[0])
+		isMotor1 := comp.Equals(cfg.Components[1])
+		isMotor2 := comp.Equals(cfg.Components[2])
+		test.That(t, isMyBase || isMotor1 || isMotor2, test.ShouldBeTrue)
+	}
+
+	// Manually inspect module
+	test.That(t, len(actualCfg.Modules), test.ShouldEqual, 1)
+	test.That(t, actualCfg.Modules[0].Equals(expectedCfg.Modules[0]), test.ShouldBeTrue)
+}
```

Comment:
Could we add one more test:
- Start a robot with a config with a modular resource (an API/model not natively supported) but no corresponding resource
- Assert that the modular resource does not build
- Reconfigure the robot to have a module that supports the resource
- Assert that the modular resource is now accessible

Reply:
good test suggestion, will add